docs/design_docs/cached_outputs.rst

```diff
@@ -0,0 +1,343 @@
+
+Requirement: unique identifier for cached outputs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Date last modified: 2021/07/30
+
+Contributors: Xylar Asay-Davis
+
+There should be a mechanism for giving each cached output file a unique 
+identifier (such as a date stamp).  A given version (git hash or release) of 
+``compass`` should know which cached files to download.  Older cached files
+should be retained so that older versions of ``compass`` can still be used
+with these cached files.  
+
+.. note::
+
+    It may be worthwhile to include a process for deprecating and then deleting
+    old cache files.
+
+.. _req_normal_or_cached:
+
+Requirement: either "normal" or "cached" versions of a step
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Date last modified: 2021/07/30
+
+Contributors: Xylar Asay-Davis
+
+We **do not** require the ability to set up a "normal" and a "cached" version
+of the same step within a ``compass`` test case or suite.  (If this is not the
+case, it would place important constraints on the design solution.)
+
+
+Design
+------
+
+.. _des_cached:
+
+Design: cached outputs
+^^^^^^^^^^^^^^^^^^^^^^
+
+Date last modified: 2021/07/30
+
+Contributors: Xylar Asay-Davis
+
+``compass`` supports "databases" of input data files on the E3SM 
+`LCRC server <https://web.lcrc.anl.gov/public/e3sm/mpas_standalonedata/>`_.
+Files will be stored in a new ``compass_cache`` database within each MPAS 
+core's space on that server.  If the "cached" version of a step is selected
+(see :ref:`des_select`), an appropriate "input" file will be added to the test 
+case where the "target" is the file on the LCRC server to be cached locally for
+future use and the "filename" is the output file.  ``compass`` will know which
+files on the server correspond to which output files via a python dictionary, 
+as described in :ref:`des_unique`.
+
+.. _des_select:
+
+Design: selecting whether to use cached outputs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Date last modified: 2021/07/30
+
+Contributors: Xylar Asay-Davis
+
+
+I haven't come up with the best way to do this yet. One possibility is that
+a ``compass`` suite could include a special line below a test case like:
+
+.. code-block:: none
+
+    ocean/global_ocean/QU240/mesh
+        cached: mesh
+    ocean/global_ocean/QU240/PHC/init
+        cached: initial_state
+
+The right mechanism for allowing users to select the cached version is even
+less clear, but it could be via a config option such as:
+
+.. code-block:: cfg
+
+    # Options related to the current test case
+    [test_case]
+    
+    # The steps from the test case where output files should be downloaded
+    # from the server rather than computed
+    cached = mesh
+
+Both of these approaches assume that we always have either the "normal" or the
+"cached" version of a step within a test case or test suite and developers or
+users are free to choose between them, see :ref:`des_normal_or_cached`.
```

Comment:
This is the main design choice where I could use some help brainstorming.  Can either of you think of better, cleaner ways of letting developers (e.g. via test suites) and users (e.g. via config files) determine which steps use cached outputs and which do not?

Reply:
I came up with a better implementation.  It's reasonably efficient for adding cached outputs to test suites and for setting up test cases where *all* steps are cached.  it's not great for setting up individual test cases where only some of the steps should be cached, and we probably just want to provide test suites if such situations emerge very often.